# Metagenomics

1. [General introduction](#General-introduction)
2. [Schedule](#Schedule)
    - [Monday](#Monday)
    - [Tuesday](#Tuesday)
    - [Wednesday](#Wednesday)
    - [Thursday](#Thursday)
    - [Friday](#Friday)
3. [Target group](#target-group)
4. [Learning outcomes](#learning-outcomes)
5. [Internet connection](#internet-connection)
7. [Organizers and teachers](#Organizers-and-teachers)


## General introduction
Microbial community structure, diversity, and population structure are fundamental aspects to understand evolution, niche adaptation and demographic history of bacterial species. As high-throughput sequencing has become cost effective and accessible, sequencing populations of bacteria across the whole metagenome provides unprecedented resolution to investigate within-host evolution, transmission history and population structure. Moreover, analysis of genetic content of microbial communities through metagenomics has become the mainstream methodology. During this one-week course you will learn and apply bioinformatic techniques to perform population genetics and study microbial communities with metagenomic approaches. The goal is to become familiar with the bioinformatic analysis tools and to be able to utilize them in your own research after the course.

**Everything is possible thanks to the support of [CSC](http://www.csc.fi)**

## Feedback
Please give us your feedback at the end of the course [here](https://presemo.helsinki.fi/metagenomics2021/).

## Schedule
The course will be held in Viikki at the Faculty of Agriculture and Forestry, University of Helsinki.  
Place: __Biocenter 1, room 1401, 1st Floor__.  
Due to COVID-19 the doors of Biocenter 1 are closed. Please call the porter to open the door (02941 59000).


### Monday  
__26.4.2021 9-16__  

| Time            | Description |
| --------------- | ----------- |
| __*Morning*__   | [Introduction](https://drive.google.com/file/d/1Fr63zkAQ8EiG7J37taBD0GMLRHaQzBu-/view?usp=sharing) (__Jenni__) |
| __*Morning*__   | [Introduction to Unix and basic usage of computing services at CSC]() (__Igor__) |
| __*Morning*__   | [Quality control and trimming of data](https://github.com/karkman/Metagenomics2021/tree/main/Day1#qc-and-trimming) hands-on, [Slides](https://drive.google.com/file/d/1qA8xxK2nwSqZo2lV3BohKQA_RlHW7QBb/view?usp=sharing) (__Jenni__)|
|                 |  **Lunch**  |
| __*Afternoon*__ | [Read based analyses] (https://drive.google.com/file/d/1EHsy1spesv3gIAALWWX7_1NBu9dwA37J/view?usp=sharing)(__Igor__) |
| __*Afternoon*__ | [QC for trimmed reads](https://github.com/karkman/Metagenomics2021/tree/main/Day1#run-qc-on-the-trimmed-reads)  |
| __*Afternoon*__ | [Read based annotation](https://github.com/karkman/Metagenomics2021/tree/main/Day1#read-based-analysis) |


### Tuesday
__21.5.2019 9-16__  

| Time            | Description |
| --------------- | ----------- |
| __*Morning*__   | [Megan interface](Day2/README.md#assembly-quality-statistics) (__Igor__)|
|                 |  **Lunch**  |
| __*Afternoon*__ | [Metaxa + R ](https://www.dropbox.com/s/yh1yf9vfuxbi9gl/2019_Helsinki_Tom_Delmont_Teaching_Material.pdf?dl=0) (__Antti__)|
| __*Afternoon*__ | [Optional: MEGAN R](Day3/README.md#02--describing-the-interface) (__Igor__)|

### Wednesday
__22.5.2019 9-16__  

| Time            | Description |
| --------------- | ----------- |
| __*Morning*__   | [Assembly lecture](Day3/README.md#02--describing-the-interface) (__Jenni__)|
| __*Morning*__   | [Assembly hands-on](https://github.com/karkman/Metagenomics2021/tree/main/Day3#assembly) (__Jenni__)|
| __*Morning*__   | [Genome resolved metagenomics](Day3/README.md#11--what-do-we-do-with-these-genomes) (__Antti__) |
|                 |  **Lunch**  |
| __*Afternoon*__ | [Preparing files for Anvi'o](Day3/README.md#11--what-do-we-do-with-these-genomes) |

### Thursday
__23.5.2019 9-16__  

| Time            | Description |
| --------------- | ----------- |
| __*Morning*__   | [Anvi’o interface](Day4#metaphlan2) (__Antti__) |
|                 |  **Lunch**  |
| __*Afternoon*__ | [Anvi’o interface](Day5#humann2) |

### Friday
__24.5.2019 9-16__  

| Time            | Description |
| --------------- | ----------- |
| 10:15-11:30     | [MAG annotation + databases lecture ](Day5#humann2) (__Igor__)|
|                 |  **Lunch**  |
| __*Afternoon*__ | [Further analyses of MAGs](https://drive.google.com/open?id=1FMBSI1df1fDFoDXud3BDMb7Z6EIST0g0) (__Igor__)|
| __*Afternoon*__ | Wrapping up, feedback, Q&A (__all__)|

## Target group
This course is targeted at PhD and MSc students who are interested in using metagenomics for investigating microbial communities.

## Learning outcomes
QA/QC of the raw sequence data, cleaning and assembly; Gene-by-gene annotation; Visualization of data; Taxonomic classification; Functional assignment, Analyses of microbial community composition; Comparative metagenomics; MAG (MetagenomeAssembledGenome) analysis; Statistical methods for metagenome data analysis.

## Internet connection
Inside the class room you can find  __eduroam__ . Instructions on how to connect in [HelpDesk website.](https://helpdesk.it.helsinki.fi/en/instructions/logging-and-connections/networks/wireless-connections-university)

## Organizers and Teachers

### Organizers
* Docent Jenni Hultman, University of Helsinki, Finland
* PhD Igor Stelmach Pessi, University of Helsinki, Finland
* PhD Antti Karkman, University of Helsinki, Finland
